### PR TITLE
interrupt properly each run through channel (and side effect fix compilation error and works on windows)

### LINKED ISSUE #63: don't rely on syscall.Kill to interrupt a run / make fortio works on Windows
I just discovered fortio doesn't run on windows:

```
C:\Users\l\go\src\istio.io>go get -u istio.io/fortio
# istio.io/fortio/ui
fortio\ui\uihandler.go:181:3: undefined: syscall.Kill
```

Ignoring for a second that go should provide a kill emulation (imo) - we probably should cleanup fortio anyway as it's only to interrupt itself

(we can also remove the exit option from the UI and only interrupt a given run or maintain a stack or list of them - or interrupt all of them through a shared channel)

### PR DESCRIPTION
**Pull request by ldemailly**
fixes #63

### CODE REVIEW

**Comment by ymesika** [comment]
@ldemailly I'm using Mac and can't verify that it indeed fixed the compilation error on Windows

**Comment by ldemailly** [comment]
@ymesika welcome back ! it's not as much about windows than about per run channel now, PTAL

File: ui/uihandler.go [hunk 1]
```diff
@@ -101,10 +104,11 @@ func (w *HTMLEscapeWriter) Write(p []byte) (int, error) {
 // Handler is the main UI handler creating the web forms and processing them.
 func Handler(w http.ResponseWriter, r *http.Request) {
 	LogRequest(r, "UI")
-	DoExit := false
-	if r.FormValue("exit") == "Exit" {
-		log.Critf("Exit request from %v", r.RemoteAddr)
-		DoExit = true
+	DoStop := false
```
Comment (ymesika) [inline ui/uihandler.go]: Seems like you can't actually get a form with both 'stop' and 'load' set together? In this case maybe DoStop and DoLoad can be combined just to hold request type.

File: ui/uihandler.go [hunk 2]
```diff
@@ -231,6 +264,9 @@ func Handler(w http.ResponseWriter, r *http.Request) {
 		if err != nil {
 			log.Errf("Unable to write json output for %v: %v", r.RemoteAddr, err)
 		}
+		if DoSave {
```
Comment (ymesika) [inline ui/uihandler.go]: Adding one more type to the request type variable I mentioned in my previous comment - stop / load / save

**Review by ymesika — APPROVED** [review]
Beside my comments below, which are cosmetics, I tested this new button and it works fine.
Wasn't able to check if it indeed got rid of the compilation error on Windows.

File: periodic/periodic.go [hunk 3]
```diff
@@ -59,14 +59,14 @@ type Runnable interface {
 
 // MakeRunners creates an array of NumThreads identical Runnable instances.
 // (for the (rare/test) cases where there is no unique state needed)
-func (ro *RunnerOptions) MakeRunners(r Runnable) {
-	log.Infof("Making %d clone of %+v", ro.NumThreads, r)
-	if len(ro.Runners) < ro.NumThreads {
-		log.Infof("Resizing runners from %d to %d", len(ro.Runners), ro.NumThreads)
-		ro.Runners = make([]Runnable, ro.NumThreads)
+func (r *RunnerOptions) MakeRunners(rr Runnable) {
```
Comment (ldemailly) [inline periodic/periodic.go:62]: this change is to minimize the diff for Normalize (as go linters want the variable name for receivers to be all the same)

File: ui/uihandler.go (hunk 2 above)
Comment (ldemailly) [reply ui/uihandler.go]: well not save as save is that checkbox is in addition to load / json

File: ui/uihandler.go (hunk 1 above)
Comment (ldemailly) [reply ui/uihandler.go]: changed to enum, thanks!

File: ui/uihandler.go [hunk 4]
```diff
@@ -155,98 +194,106 @@ func Handler(w http.ResponseWriter, r *http.Request) {
...
+			for _, v := range runs {
+				v.Abort()
+				i++
+			}
+			uiRunMapMutex.Unlock()
+			log.Infof("Interrupted %d runs", i)
+		} else { // Stop one
+			uiRunMapMutex.Lock()
+			v := runs[runid]
+			if v != nil {
```
Comment (rkpagadala) [inline ui/uihandler.go]: Change to
v, ok := runs[runid]
 			if ok {

File: periodic/periodic.go [hunk 5]
```diff
@@ -128,9 +132,15 @@ type periodicRunner struct {
 	RunnerOptions
 }
 
-// internal version, returning the concrete implementation.
-func newPeriodicRunner(opts *RunnerOptions) *periodicRunner {
-	r := &periodicRunner{*opts} // by default just copy the input params
+var (
+	globalAbort     chan os.Signal
```
Comment (rkpagadala) [inline periodic/periodic.go]: Consider changing to
gAbortChan

File: periodic/periodic.go [hunk 6]
```diff
@@ -128,9 +132,15 @@ type periodicRunner struct {
...
+	outstandingRuns int64
```
Comment (rkpagadala) [inline periodic/periodic.go]: gOutstandingRuns

File: periodic/periodic.go [hunk 7]
```diff
@@ -128,9 +132,15 @@ type periodicRunner struct {
...
+	abortMutex      sync.Mutex
```
Comment (rkpagadala) [inline periodic/periodic.go]: gAbortMutex

Or put then all in a structure and call it gSomething

**Review by rkpagadala — APPROVED** [review]
Minor comments

File: periodic/periodic.go (hunk 5 above)
Comment (ldemailly) [reply periodic/periodic.go]: k

File: periodic/periodic.go (hunk 6 above)
Comment (ldemailly) [reply periodic/periodic.go]: k

File: periodic/periodic.go (hunk 7 above)
Comment (ldemailly) [reply periodic/periodic.go]: k

File: ui/uihandler.go (hunk 4 above)
Comment (ldemailly) [reply ui/uihandler.go]: k

**Pull request merged**